service/qos/service_level_controller.hh

```diff
@@ -67,6 +78,7 @@ private:
         semaphore notifications_serializer = semaphore(1);
         future<> distributed_data_update = make_ready_future();
         abort_source dist_data_update_aborter;
+        abort_source group0_aborter;
```

Comment:
why it has 2 abort sources? is it used in any other case than shutdown? if we are shutting down we need to abort everything so it'd seem one source per class would be enough

Reply:
We use it to abort group0 operations. The SL controller has longer lifetime than group0; when group0 is being shut down, we must first make sure to abort outstanding group0 operations in SL controller first.

We had a similar issue in storage_service and it was solved in a similar way: https://github.com/scylladb/scylladb/commit/8ed8b151da732260e2d6322839cffb5b497103ab

Reply:
ok. Wouldn't be better then to "subscribe" to group0 lifetime more directly? Something for the future to think about.